ESQL: Add more time span units

bpintea opened the pull request:
This adds ~`nanosecond`, `microsecond` and~ `quarter` to the set of supported time spans. It also adds a few standard and common abbreviations to some existing ones.

Edit: not adding sub-millisecond units, since these aren't currently supported (resolution currently support millis only).

bpintea committed:
Add more time span units

This adds nanosecond, microsecond and quarter to the set of supported
time spans. It also adds a few standard and common abbreviations to
some existing ones.

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

Commit:
Update docs/changelog/108300.yaml

costin commented on docs/reference/esql/esql-syntax.asciidoc:
```diff
@@ -160,14 +160,17 @@ Datetime intervals and timespans can be expressed using timespan literals.
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
 * `week`/`weeks`
```
How about `w` as a shorthand for weeks? There doesn't seem to be any ambiguity (especially not around the near units).

bpintea replied:
I've added a `q` for quarter too, btw.

costin commented on docs/reference/esql/esql-syntax.asciidoc:
```diff
@@ -160,14 +160,17 @@ Datetime intervals and timespans can be expressed using timespan literals.
-* `year`/`years`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`
+* `year`/`years`/`yr`
```
How about `y` : `now() - 1y`? Doesn't seem to be any other y unit near it.

alex-spies replied:
++

costin reviewed (approved):
LGTM

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java:
```diff
@@ -373,32 +373,47 @@ public void testUnquotedIdentifiers() {
     public void testDurationLiterals() {
         int value = randomInt(Integer.MAX_VALUE);
```
Super nit: Maybe we should use a random long instead to better cover the relevant range (esp. for nanos).

bpintea replied:
IIRC some of the static Duration/Period methods fail with too large values and this was the reason to use an int here.

alex-spies reviewed (commented):
Nice addition! But I think we need to add more tests.
- `AbstractFunctionTestCase.randomLiteral()` never emits `time_duration`s in units smaller than millis. The same applies to `TestCaseSupplier.timeDurationCases`.
- None of our csv tests use the new units, so there's no "proof" that they actually work end-to-end, resp. nothing guarding against regressions. At least, I think we should have some date arithmetic tests for `datetime +/- x` where `x` is given in nanos, mus, and quarters.

bpintea committed:
Remove sub-millisecond spans

bpintea committed:
Merge branch 'main' into esql/more_time_span

bpintea commented:
> AbstractFunctionTestCase.randomLiteral() never emits time_durations in units smaller than millis. The same applies to TestCaseSupplier.timeDurationCases.

And they should stay the same, since it wouldn't make a difference, as we don't support nanos. Which means it makes no sense adding them a.t.p. -- I've removed them. 🙏 

> At least, I think we should have some date arithmetic tests for datetime +/- x where x is given in nanos, mus, and quarters.

I've added some csv-spec tests with quarter and the new abbreviations. I think the unit tests would have been relevant for parsing these units, but more tests is better than less tests.
(Btw, csv-spec tests with sub-millisecond units would have worked, but these wouldn't have made a difference to the outcome.)

bpintea committed:
disable some bwc versions

bpintea committed:
disable some bwc, but with features

bpintea committed:
Introduce new NodeFeature

alex-spies commented on docs/reference/esql/esql-syntax.asciidoc:
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
-* `week`/`weeks`
-* `month`/`months`
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`s`
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
+* `week`/`weeks`/`w`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`/`q`
+* `year`/`years`/`yr`/`y`
```
- Are we sure we want to deviate from the [standard abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/8.13/common-options.html#date-math) used in ES? Applies to `mo` and `min`. In SQL, apparently there is yet another [set of abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-add); not sure if we should follow suit, but having 3 different sets of abbreviations will likely be confusing.
- Are we sure that the time period `min` and the aggregation function `min` will never be ambiguous when parsing?

bpintea replied:
> In SQL, apparently there is yet another [set of abbreviations](https://www.elastic.co/guide/en/elasticsearch/reference/current/sql-functions-datetime.html#sql-functions-datetime-add); not sure if we should follow suit, but having 3 different sets of abbreviations will likely be confusing.

Interestingly that you've made this comparison too. I've recently added [some clarifications](https://github.com/elastic/elasticsearch/pull/108301) about the non-overlapping of this set with function parameters.

> Are we sure that the time period min and the aggregation function min will never be ambiguous when parsing?

We have the tests to keep us in check. But I also hope we won't need to change the grammar in a way that there might be conflicts.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/type/EsqlDataTypeConverter.java:
```diff
@@ -234,18 +234,20 @@ public static DataType commonType(DataType left, DataType right) {
         return DataTypeConverter.commonType(left, right);
     }
 
+    // generally supporting abbreviations from https://en.wikipedia.org/wiki/Unit_of_time
```
Then we should also add smaller time units, like one [Jiffy](https://en.wikipedia.org/wiki/Unit_of_time#Scientific) and the Planck time. /s

alex-spies commented on docs/reference/esql/esql-syntax.asciidoc:
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
-* `month`/`months`
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`s`
```
Missing `sec`
```suggestion
* `second`/`seconds`/`sec`/`s`
```

bpintea replied:
Yeh... I guess if we opened the gates, "sec" can be there too. I'll add it.

alex-spies commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/parser/ExpressionTests.java:
```diff
@@ -380,37 +380,51 @@ public void testDurationLiterals() {
         assertEquals(l(Duration.ZERO, TIME_DURATION), whereExpression("0 second"));
         assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + "second"));
         assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + " seconds"));
+        assertEquals(l(Duration.ofSeconds(value), TIME_DURATION), whereExpression(value + " s"));
```
Missing `sec`.

bpintea replied:
Thx!

bpintea committed:
review comments

bpintea committed:
Merge branch 'main' into esql/more_time_span

bpintea commented:
Thx Costin, Alex.

drewdaemon commented on docs/reference/esql/esql-syntax.asciidoc:
```diff
@@ -160,14 +160,15 @@ Datetime intervals and timespans can be expressed using timespan literals.
-* `week`/`weeks`
-* `month`/`months`
-* `year`/`years`
+* `millisecond`/`milliseconds`/`ms`
+* `second`/`seconds`/`sec`/`s`
+* `minute`/`minutes`/`min`
+* `hour`/`hours`/`h`
+* `day`/`days`/`d`
+* `week`/`weeks`/`w`
+* `month`/`months`/`mo`
+* `quarter`/`quarters`/`q`
+* `year`/`years`/`yr`/`y`
```
Looks like we missed `microsecond` and `nanosecond` cc @bpintea

bpintea replied:
I've updated the PR description. We don't yet support sub-millisecond values.
